libmtd/mtd.c

```diff
@@ -0,0 +1,529 @@
+		ret = -EINVAL;
+	}
+	else if (mtd->ops->isLocked == NULL) {
+		ret = -EOPNOTSUPP;
+	}
+	else if (len == 0) {
+		ret = EOK;
+	}
+	else {
+		ret = mtd->ops->isLocked(mtdInfo->storage, ofs, len);
+	}
+
+	return ret;
+}
+
+
+int mtd_block_isreserved(struct mtd_info *mtdInfo, off_t ofs)
+{
+	int ret;
+	storage_mtd_t *mtd = mtdInfo->storage->dev->mtd;
+
+	if (ofs < 0 || ofs >= mtdInfo->size) {
+		ret = -EINVAL;
+	}
+	else if (mtd->ops->block_isReserved == NULL) {
+		ret = -EOPNOTSUPP;
+	}
+	else {
+		ret = mtd->ops->block_isReserved(mtdInfo->storage, ofs);
+	}
+
+	return ret;
+}
+
+
+/* TODO: test NAND flash memory */
+int mtd_block_isbad(struct mtd_info *mtdInfo, off_t ofs)
+{
+	int ret;
+	storage_mtd_t *mtd = mtdInfo->storage->dev->mtd;
+
+	if (ofs < 0 || ofs >= mtdInfo->size) {
+		ret = -EINVAL;
+	}
+	else if (mtd->ops->block_isBad == NULL) {
+		ret = -EOPNOTSUPP;
+	}
+	else {
+		ret = mtd->ops->block_isBad(mtdInfo->storage, ofs);
+	}
+
+	return ret;
+}
+
+
+/* TODO: test NAND flash memory */
+int mtd_block_markbad(struct mtd_info *mtdInfo, off_t ofs)
+{
+	int ret;
+	storage_mtd_t *mtd = mtdInfo->storage->dev->mtd;
+
+	if (ofs < 0 || ofs >= mtdInfo->size) {
+		ret = -EINVAL;
+	}
+	else if (mtd->ops->block_markBad == NULL) {
+		ret = EOK;
+	}
+	else {
+		ret = mtd->ops->block_markBad(mtdInfo->storage, ofs);
+	}
+
+	return ret;
+}
+
+
+int mtd_suspend(struct mtd_info *mtdInfo)
+{
+	int ret;
+	storage_mtd_t *mtd = mtdInfo->storage->dev->mtd;
+
+	if (mtd->ops->suspend == NULL) {
+		ret = -EOPNOTSUPP;
+	}
+	else {
+		ret = mtd->ops->suspend(mtdInfo->storage);
+	}
+
+	return ret;
+}
+
+
+void mtd_resume(struct mtd_info *mtdInfo)
+{
+	storage_mtd_t *mtd = mtdInfo->storage->dev->mtd;
+
+	if (mtd->ops->resume != NULL) {
+		mtd->ops->resume(mtdInfo->storage);
+	}
+}
+
+
+static void *mtd_allocAligned(size_t size, size_t alignment)
+{
+	addr_t pa;
+	char *addr, *res;
+
+	size = (size + (_PAGE_SIZE - 1)) & ~(_PAGE_SIZE - 1);
+	alignment = (alignment + (_PAGE_SIZE - 1)) & ~(_PAGE_SIZE - 1);
+
+	addr = mmap(NULL, size, PROT_WRITE | PROT_READ, MAP_ANONYMOUS, OID_CONTIGUOUS, 0);
+	if (addr == MAP_FAILED) {
+		return NULL;
+	}
+
+	pa = va2pa(addr);
+	if ((pa % alignment) == 0) {
+		return addr;
+	}
+
+	munmap(addr, size);
+
+	addr = mmap(NULL, size, PROT_WRITE | PROT_READ, MAP_ANONYMOUS, OID_CONTIGUOUS, 0);
+	if (addr == MAP_FAILED) {
+		return NULL;
+	}
+
+	res = (char *)(((uintptr_t)addr + (alignment - 1)) & ~(alignment - 1));
+
+	/* unmap memory prefix */
+	if (res != addr) {
+		munmap(addr, res - addr);
+	}
+
+	/* unmap memory sufix */
+	if ((res + size) != (addr + alignment * 2)) {
+		munmap(res + size, (addr + alignment * 2) - (res + size));
+	}
+
+	return res;
+}
+
+
+void *mtd_kmalloc_up_to(const struct mtd_info *mtdInfo, size_t *size)
+{
+	void *ret;
+
+	if (*size < mtdInfo->writesize) {
+		*size = mtdInfo->writesize;
+	}
+
+	if (mtdInfo->writesize < _PAGE_SIZE) {
+		ret = mmap(NULL, *size, PROT_READ | PROT_WRITE, MAP_ANONYMOUS, OID_CONTIGUOUS, 0);
```

Comment:
Won't work for `*size != n * _PAGE_SIZE`. We need to align `*size` both to `mtdInfo->writesize` and `_PAGE_SIZE`.